---
title: "Data normalization and pre-processing: Monocytes/Macrophages"
author: "Laura Jiménez Gracia"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    toc: yes
    toc_float: yes
    number_sections: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE,
                      fig.align = "center")
```

# Introduction
Following with the results obtained in the previous notebook, we will now perform other pre-processing steps, such as data normalization, feature selection, scaling, dimensionality reduction, and data visualization prior to batch-effect correction (data integration). To do so, we will follow the current best practices for scRNA-seq QC described in [Luecken al. Mol Syst Biol (2018)](https://doi.org/10.15252/msb.20188746) and adapt some workflows from [Satija Lab](https://satijalab.org/seurat/vignettes.html).

## Objective
In this Rmarkdown document, we are going to perform the previous pre-processing steps.

# Pre-processing
## Libraries
```{r warning = FALSE, message = FALSE}
library(tidyverse)
library(Seurat)
library(gt)
library(dittoSeq)
library(harmony)
```

## Parameters
Here we will define and load objects and functions that will be used throughout the document.
```{r}
# Paths
path_project_metadata <- here::here("01_cellranger_mapping/data/POST_12_metadata.csv")
path_samples_metadata <- here::here("01_cellranger_mapping/data/POST_12_metadata_samples_all.csv")
path_r_objects <- here::here("03_clustering_annotation/results/R_objects")
path_r_tables <- here::here("03_clustering_annotation/results/tables")
path_r_figs <- here::here("03_clustering_annotation/results/figs")

# Functions
source(here::here("bin/utils.R"))

# Parameters
confounder_variables <- c("library_name", "sample_id", "genotype", "treatment")
confounder_names <- c("Library", "Sample", "Genotype", "Treatment")

metadata_variables_confounder <- c("nCount_RNA", "nFeature_RNA", "gem_id", 
                                   "pct_mt", "pct_rb", "Phase",
                                   "library_name", "sample_id", "genotype", "treatment")

color_palette = c("aquamarine", "orange", "lightgreen", "blueviolet", "maroon",
                  "deepskyblue3", "darkgray", "red", "green", "gold3", "magenta", 
                  "pink", "lightgray", "blue", "darkgreen",  "gold", "coral2", 
                  "pink3", "yellow", "palegreen", "peachpuff", "steelblue", "cyan3",
                  "darkblue", "lightblue", "green3", "aquamarine", "orange",
                  "darkgray", "red", "green", "deepskyblue3",  "gold3", "magenta", 
                  "pink", "violet", "blue", "darkgreen",  "gold", "coral2", 
                  "navy", "yellow", "darkblue", "palegreen", "peachpuff", "steelblue",
                  "cyan3", "lightblue", "green3")
```

## Load data
The data used in this Rmarkdown document comes from **POST_12** dataset.
```{r}
# Load all metadata information
metadata <- read.csv(path_project_metadata)

# Merging metadata_all tables
metadata <- metadata %>% filter(type == "cDNA" & fraction == "CD45+")

print("Libraries/Samples metadata")
DT::datatable(metadata, options = list(scrollX = TRUE))

# Load Seurat object
seurat_obj <- readRDS(paste0(path_r_objects, "/post12_cd45pos_monomacro.rds"))
seurat_obj
```

## Explore QC results
Here, we show an overview of the scRNA-seq data obtained after the quality control.
```{r}
table_qc_gex(seurat_obj@meta.data, subtitle = "After cell QC filtering out")
```

```{r fig.height=10, fig.width=14}
gg_qc_by_gemid <- VlnPlot(seurat_obj,
                          features = c("nCount_RNA", "nFeature_RNA",
                                       "pct_mt", "pct_rb"),
                          group.by = "sample_id",
                          pt.size = 0,
                          ncol = 2
                          )

gg_qc_by_gemid
```

# Data normalization
Count depths for identical cells can be different due to the variability inherent in the capture, RT and sequencing of a mRNA molecule. Thus, to avoid detecting gene expression differences based on count data when comparing cells we will apply a global-scaling normalization by sequencing depth. 

This method divides the feature expression measurements (each count) by the library size of the cell (total number of UMIs), multiplies this by a scale factor (10,000) and log-transforms it.

```{r}
# Normalization
seurat_obj <- NormalizeData(
  seurat_obj,
  normalization.method = "LogNormalize",
  scale.factor = 10000)

seurat_obj[["RNA"]]@data[1:10, 1:30]
```


# Feature selection and data scaling
Next, we will filter a subset of features that exhibit high cell-to-cell variation to keep only highly variable genes that are “informative” of the variability in the dataset, which will be further used in downstream analysis.

```{r}
# Summary statistics of the number of features (genes) by cell
summary(Matrix::colSums(seurat_obj@assays$RNA@data[, ] > 0))
```

```{r fig.width=7, fig.height=5}
# Identify highly variable features
seurat_obj <- FindVariableFeatures(
  seurat_obj,
  nfeatures = 2000)

# Identify the 20 most highly variable genes
top20_hvf <- head(VariableFeatures(seurat_obj), 20)
top20_hvf

# Plot variable features
LabelPoints(VariableFeaturePlot(seurat_obj),
            points = top20_hvf,
            repel = TRUE)
```

Moreover, we will also apply a linear transformation `scaling` prior to dimensional reduction to avoid highly-expressed genes to dominate in the analysis. The scaling step shifts the expression of each gene, so that the mean expression across cells is 0 and the variance is 1.
```{r}
# Scaling
seurat_obj <- ScaleData(seurat_obj)
```

# Dimensionality reduction
Next, we will perform a linear dimensionality reduction using the `PCA` algorithm to reduce the high-dimension of data (we have many genes, and some of them are redundant).
```{r fig.width=7, fig.height=5}
# Run PCA
seurat_obj <- RunPCA(
  seurat_obj,
  ndims.print = 1:6,
  nfeatures.print = 10)

# Visualizing PCA results
# VizDimLoadings(seurat_obj, dims = 1:2, reduction = "pca")
```

After applying PCA, we will ensure that the first PCs capture cell-to-cell variability and not differences associated to batch-effects or by cell cycle phase.
```{r fig.width=15, fig.height=5}
gg_pca_by_lib <- DimPlot(
  seurat_obj,
  reduction = "pca",
  group.by = "library_name") +
  labs(title = "", color = "Library")

gg_pca_by_sample <- DimPlot(
  seurat_obj,
  reduction = "pca",
  group.by = "sample_id") +
  labs(title = "", color = "Sample")

gg_pca_by_lib + gg_pca_by_sample
```


```{r fig.width=12, fig.height=5}
# Visualize the PCA reduction, grouping by cell cycle phase
gg_pca_by_cellcycle <- DimPlot(
  seurat_obj,
  reduction = "pca",
  group.by = "Phase",
  split.by = "Phase",
  pt.size = 0.01) + NoLegend()

gg_pca_by_cellcycle
```
Based on the previous plots, we do not observe significant batch-effects neither cell cycle phase effects. Thus, we would not regress out the variation due to cell cycle because there is no effect.

At this point, we will determine how many PCs should be considered for data visualization in 2 dimensions and for the downstream cluster analysis. It important to ensure that we are capturing the majority of the variation (or cell types) present in our dataset.

```{r fig.width=7, fig.height=5}
# Elbowplot to select relevant PCs to analyse the dataset
ElbowPlot(seurat_obj, ndims = 50, reduction = "pca")
```

In this particular dataset, we will use the first 20 PCs to visualize data and perform the clustering.

# Data visualization
For visualization purposes, it is standard practice to use the Uniform Approximation and Projection method (UMAP), which is a non-linear dimensionality reduction method that reduce the data dimensionality into a 2-dimension space.

```{r}
# Run UMAP
seurat_obj <- seurat_obj %>% 
  RunUMAP(reduction = "pca", dims = 1:20)
```

## Assessing confounder variables

```{r fig.width=14, fig.height=10}
# Visualizing UMAP
gg_umap_uncorrected <- purrr::map2(confounder_variables, confounder_names,
                                          function(confounder, title) {
  p <- DimPlot(seurat_obj, 
               reduction = "umap",
               group.by = confounder,
               pt.size = 0.01) +
    labs(title=title) +
    scale_color_brewer(palette = "Paired")
  p
})

cowplot::plot_grid(plotlist = gg_umap_uncorrected, ncol = 2)
```

Here, from the UMAP mapping, we can see the presence of library/donor-specific effects that should not be present. Thus, we will integrate different libraries to overcome the batch effect.

## Assessing cell cycle effects
```{r fig.width=7, fig.height=5}
gg_umap_by_phase <- DimPlot(
  seurat_obj,
  reduction = "umap",
  group.by = "Phase",
  pt.size = 0.01) +
  labs(color = "Cell Cycle Phase")

gg_umap_by_phase
```
Finally, we ensure that there is no batch effect associated to the cell cycle phase.


## Assessing sources of uninterested variation
Next we will explore additional metrics, such as the number of UMIs and genes per cell, S-phase and G2M-phase markers, and mitochondrial / ribosomal gene expression on the UMAP embedding. Finally, we will also check the scrublet doublet score.
```{r fig.width=14, fig.height=15}
# Determine metrics to plot present in seurat_integrated@meta.data
qc_metrics <-  c("nCount_RNA", "nFeature_RNA", 
                 "S.Score", "G2M.Score",
                 "pct_mt", "pct_rb")

gg_qc_umap <- FeaturePlot(
  seurat_obj,
  reduction = "umap",
  features = qc_metrics,
  order = TRUE,
  ncol = 2)

gg_qc_umap
```

## Assessing multiple confounder variables
```{r fig.width=14, fig.height=5}
DimPlot(
  seurat_obj,
  group.by = "sample_id",
  split.by = "library_name",
  pt.size = 0.01
  ) +
  labs(title = "Library",
       color = "Sample") +
  scale_color_brewer(palette = "Paired")

DimPlot(
  seurat_obj,
  group.by = "treatment",
  split.by = "genotype",
  pt.size = 0.01
  ) +
  labs(title = "Genotype",
       color = "Treatment/Disease") +
  scale_color_brewer(palette = "Paired")
```


```{r fig.width=20, fig.height=5}
DimPlot(
  seurat_obj,
  group.by = "treatment",
  split.by = "sample_id",
  pt.size = 0.01
  ) +
  labs(title = "Sample",
       color = "Treatment/Disease") +
  scale_color_brewer(palette = "Paired")
```


# Integration with harmony
The integration approach was carried out with [Harmony](https://www.nature.com/articles/s41592-019-0619-0). [Chazarra-Gil *et al*](https://www.biorxiv.org/content/10.1101/2020.05.22.111211v2), which was reported to be among the best scRNA-seq integration methods, especially in terms of scalability, in a recent integration benchmarking study.
```{r}
seurat_obj_integrated <- seurat_obj %>%
  RunHarmony(group.by.vars = c("sample_id"))

# Elbowplot to select relevant PCs to analyse the dataset
ElbowPlot(seurat_obj_integrated, ndims = 50, reduction = "harmony")

# Dimensionality reduction
seurat_obj_integrated <- seurat_obj_integrated %>%
  RunUMAP(reduction = "harmony", dims = 1:20)
```


## Assessing confounder variables
```{r fig.width=14, fig.height=10}
# Visualizing UMAP
gg_umap_seurat_confounders <- purrr::map2(confounder_variables, confounder_names, 
                                                 function(confounder, title) {
  gg <- DimPlot(seurat_obj_integrated, 
               reduction = "umap",
               group.by = confounder,
               pt.size = 0.1) +
    labs(title=title) +
    scale_color_brewer(palette = "Paired")
  gg
})

cowplot::plot_grid(plotlist = gg_umap_seurat_confounders, ncol = 2)
```

## Assessing cell cycle effects
```{r fig.width=7, fig.height=5}
gg_umap_by_phase <- DimPlot(
  seurat_obj_integrated,
  reduction = "umap",
  group.by = "Phase",
  pt.size = 0.01) +
  labs(color = "Cell Cycle Phase")

gg_umap_by_phase
```

## Assessing sources of uninterested variation
Next we will explore additional metrics, such as the number of UMIs and genes per cell, S-phase and G2M-phase markers, and mitochondrial / ribosomal gene expression on the UMAP embedding. Finally, we will also check the scrublet doublet score.
```{r fig.width=14, fig.height=15}
# Determine metrics to plot present in seurat_integrated@meta.data
qc_metrics <-  c("nCount_RNA", "nFeature_RNA", 
                 "S.Score", "G2M.Score",
                 "pct_mt", "pct_rb")

gg_umap_integrated_qc <- FeaturePlot(
  seurat_obj_integrated,
  reduction = "umap",
  features = qc_metrics,
  order = TRUE,
  ncol = 2)

gg_umap_integrated_qc
```

## Assessing multiple confounder variables
```{r fig.width=14, fig.height=5}
DimPlot(
  seurat_obj_integrated,
  group.by = "sample_id",
  split.by = "library_name",
  pt.size = 0.01
  ) +
  labs(title = "Library",
       color = "Sample") +
  scale_color_brewer(palette = "Paired")

DimPlot(
  seurat_obj_integrated,
  group.by = "treatment",
  split.by = "genotype",
  pt.size = 0.01
  ) +
  labs(title = "Genotype",
       color = "Treatment/Disease") +
  scale_color_brewer(palette = "Paired")
```


```{r fig.width=20, fig.height=5}
DimPlot(
  seurat_obj_integrated,
  group.by = "treatment",
  split.by = "sample_id",
  pt.size = 0.01
  ) +
  labs(title = "Sample",
       color = "Treatment/Disease") +
  scale_color_brewer(palette = "Paired")
```

# Clustering
```{r}
## Define resolutions to compute
resolutions_range <- c(0.01, 0.05, 0.1, 0.2, 0.25,
                       0.3, 0.4, 0.5, 0.6, 0.7, 
                       0.75, 0.8, 0.9, 1, 1.1, 
                       1.2, 1.25, 1.3, 1.4, 1.5)
```


```{r fig.width=25, fig.height=25}
resolution_names <- as.character(purrr::map(resolutions_range, function(res) {
  paste0("RNA_snn_res.", res)}))

# Clustering
## Determine the K-nearest neighbor graph
seurat_obj_integrated <- FindNeighbors(
  seurat_obj_integrated,
  reduction = "harmony",
  dims = 1:20)

## Determine the clusters (Louvain algorithm) for multiple resolutions                                
seurat_obj_integrated <- FindClusters(
  seurat_obj_integrated,
  resolution = resolutions_range,
  verbose = FALSE)


# Clustering overview
gg_umap_cluster_resolution <- DimPlot(object = seurat_obj_integrated,
                                      group.by = resolution_names,
                                      label = TRUE,
                                      label.size = 3,
                                      cols = color_palette,
                                      ncol = 4
                                      ) & NoLegend()
gg_umap_cluster_resolution
```


## Explore cell clusters
```{r}
## Select cluster resolution
Idents(seurat_obj_integrated) <- "RNA_snn_res.0.5"
```

```{r}
DimPlot(object = seurat_obj_integrated,
        label = TRUE,
        label.size = 6,
        repel = TRUE,
        cols = color_palette
        ) & NoLegend()
```


Now, we will also explore clusters across sample.
```{r fig.width=20, fig.height=5}
DimPlot(
  object = seurat_obj_integrated,
  split.by = "sample_id",
  label = FALSE,
  cols = color_palette
  )
```


```{r}
# Number of cells / cluster
table(seurat_obj_integrated@active.ident)

# Number of cells per cluster vs confounder variables
table(seurat_obj_integrated$sample_id, seurat_obj_integrated@active.ident)
table(seurat_obj_integrated$treatment, seurat_obj_integrated@active.ident)
```

## Find Markers
```{r}
# Define Cluster biomarkers
## Find diferentially expressed features
seurat_obj_markers <- seurat_obj_integrated %>% 
  FindAllMarkers(
    only.pos = TRUE,
    min.pct = 0.25,
    logfc.threshold = 0.25,
    test.use = "wilcox")

seurat_obj_markers <- seurat_obj_markers %>%
  dplyr::filter(p_val_adj < 0.05) 

# Save cell-type biomarkers
## Export in .xlsx format
### Prepare data: sort and filter
biomarkers_df <- seurat_obj_markers %>%
  dplyr::arrange(cluster, desc(abs(avg_log2FC))) %>%
  dplyr::filter(avg_log2FC > 0.5)

### Convert biomarkers dataframe into a biomarker list by cluster
biomarkers_list <- purrr::map(levels(biomarkers_df$cluster),
                              ~ biomarkers_df[biomarkers_df$cluster == .x, ])
names(biomarkers_list) <- levels(biomarkers_df$cluster)

openxlsx::write.xlsx(biomarkers_list,
                     file = paste0(path_r_tables, "/biomarkers_cd45pos_monomacro_resolution-0.5.xlsx"))
```

## Explore cluster biomarkers
### Table
```{r}
# Visualize top markers per cluster
top_all_markers <- seurat_obj_markers %>%
  group_by(cluster) %>%
  slice_max(n = 10, order_by = avg_log2FC)

top_all_markers
DT::datatable(top_all_markers, options = list(scrollX = TRUE))
```

### Dotplot
```{r fig.width=20, fig.height=5}
# Visualize top markers per cluster
top_all_markers <- seurat_obj_markers %>%
  group_by(cluster) %>%
  slice_max(n = 4, order_by = avg_log2FC)

top_genemarkers_list <- unique(top_all_markers$gene)

DotPlot(
  seurat_obj_integrated,
  features = top_genemarkers_list,
  dot.scale = 6) +
  RotatedAxis()
```


### Explore gene markers

```{r fig.width=20, fig.height=5}
mono_genes <- c("Plac8", "Ly6c2", "Chil3", "Hp")

FeaturePlot(seurat_obj_integrated,
            features = mono_genes,
            label = T,
            #split.by = "tissue",
            order = F,
            pt.size = 0.25,
            ncol = 4
)

VlnPlot(seurat_obj_integrated,
            features = mono_genes,
            #split.by = "tissue",
            #order = T,
            pt.size = 0.25,
            ncol = 4
)
```



```{r fig.width=20, fig.height=5}
monomat_genes <- c("Rsad2", "Irf7", "Ms4a4c", "Isg15")

FeaturePlot(seurat_obj_integrated,
            features = monomat_genes,
            label = T,
            #split.by = "tissue",
            order = F,
            pt.size = 0.25,
            ncol = 4
)

VlnPlot(seurat_obj_integrated,
            features = monomat_genes,
            #split.by = "tissue",
            #order = T,
            pt.size = 0.25,
            ncol = 4
)
```


```{r fig.width=20, fig.height=5}
monoDC_genes <- c("Plac8", "Fabp4", "Fcgr4", "Itgal", "Ear2")

FeaturePlot(seurat_obj_integrated,
            features = monoDC_genes,
            label = T,
            #split.by = "tissue",
            order = F,
            pt.size = 0.25,
            ncol = 5
)

VlnPlot(seurat_obj_integrated,
            features = monoDC_genes,
            #split.by = "tissue",
            #order = T,
            pt.size = 0.25,
            ncol = 5
        
)
```

```{r fig.width=15, fig.height=10}
M1macro_genes <- c("Ifitm1", "Thbs1", "Flrt3", "Il7r", "Il1b", "Ccr1")

FeaturePlot(seurat_obj_integrated,
            features = M1macro_genes,
            label = T,
            #split.by = "sample_id",
            order = F,
            pt.size = 0.25,
            ncol = 3
)

VlnPlot(seurat_obj_integrated,
            features = M1macro_genes,
            #split.by = "tissue",
            #order = T,
            pt.size = 0.25,
            ncol = 3
)
```



```{r fig.width=20, fig.height=5}
Slamf9macro_genes <- c("Slamf9", "Cd72", "Itgam", "Adgre1")

FeaturePlot(seurat_obj_integrated,
            features = Slamf9macro_genes,
            label = T,
            #split.by = "sample_id",
            order = F,
            pt.size = 0.25,
            ncol = 4
)

VlnPlot(seurat_obj_integrated,
            features = Slamf9macro_genes,
            #split.by = "sample_id",
            pt.size = 0.25,
            ncol = 4
)
```


```{r fig.width=25, fig.height=10}
M2macro_genes <- c("Apoe", "Selenop", "C1qa", "C1qb", "Cd163", "Cd68", "Mrc1", "Cd86", "Adgre1", "Cd300c2")

FeaturePlot(seurat_obj_integrated,
            features = M2macro_genes,
            label = T,
            #split.by = "tissue",
            order = F,
            pt.size = 0.25,
            ncol = 5
)

VlnPlot(seurat_obj_integrated,
            features = M2macro_genes,
            #split.by = "tissue",
            #order = T,
            pt.size = 0.25,
            ncol = 5
)
```

```{r fig.width=20, fig.height=10}
MHCmacro_genes <- c("Cd74", "H2-DMa", "H2-DMb1", "H2-Ab1", "H2-Aa", "H2-Eb1", "Ccr2", "Clec4b1")

FeaturePlot(seurat_obj_integrated,
            features = MHCmacro_genes,
            label = T,
            #split.by = "sample_id",
            order = F,
            pt.size = 0.25,
            ncol = 4
)

VlnPlot(seurat_obj_integrated,
            features = MHCmacro_genes,
            #split.by = "sample_id",
            #order = T,
            pt.size = 0.25,
            ncol = 4
)
```


```{r fig.width=20, fig.height=15}
dermalmacro_genes <- c("Cd209d", "Cd163", "Folr2", "Cxcl13")

FeaturePlot(seurat_obj_integrated,
            features = dermalmacro_genes,
            label = T,
            #split.by = "sample_id",
            order = F,
            pt.size = 0.25,
            ncol = 4
)

VlnPlot(seurat_obj_integrated,
            features = dermalmacro_genes,
            #split.by = "sample_id",
            #order = T,
            pt.size = 0.25,
            ncol = 4
)
```


Exploring gene expression based on https://www.science.org/doi/10.1126/sciimmunol.abf7777
```{r fig.width=20, fig.height=5}
four_genes <- c("Timd4", "Lyve1", "Folr2", "Ccr2")

FeaturePlot(seurat_obj_integrated,
            features = four_genes,
            label = T,
            #split.by = "sample_id",
            #order = T,
            pt.size = 0.25,
            ncol = 4
)

VlnPlot(seurat_obj_integrated,
            features = four_genes,
            #split.by = "sample_id",
            #order = T,
            pt.size = 0.25,
            ncol = 4
)
```

```{r fig.width=30, fig.height=10}
TLFmacro_genes <- c("Timd4", "Lyve1", "Folr2", "Ninj1", "Gas6", "Mrc1", "Cd163", "Igf1", "Selenop", "Igfbp4", "F13a1")
# Selenop == Sepp1, 

FeaturePlot(seurat_obj_integrated,
            features = TLFmacro_genes,
            label = T,
            #split.by = "sample_id",
            #order = T,
            pt.size = 0.25,
            ncol = 6
)

VlnPlot(seurat_obj_integrated,
            features = TLFmacro_genes,
            #split.by = "sample_id",
            #order = T,
            pt.size = 0.25,
            ncol = 6
)
```



```{r fig.width=25, fig.height=15}
CCR2macro_genes <- c("Ccr2", "Cd52", "H2-Eb1", "H2-Aa", "H2-Ab1", "Bcl2a1a","S100a4", "S100a6", "S100a8", "S100a11", "Coro1a", "Cyp4f18", "Fxdy5")


FeaturePlot(seurat_obj_integrated,
            features = CCR2macro_genes,
            label = T,
            #split.by = "sample_id",
            #order = T,
            pt.size = 0.25,
            ncol = 5
)

VlnPlot(seurat_obj_integrated,
            features = CCR2macro_genes,
            #split.by = "sample_id",
            #order = T,
            pt.size = 0.25,
            ncol = 5
)
```


```{r fig.width=30, fig.height=10}
MHCmacro_genes <- c("Cd14", "Cst3", "Lilra5", "Apoe", "Trem2", "H3f3b", "C5ar1", "St3gal6", "Stab1", "H2-Eb1", "H2-Aa", "H2-Ab1")


FeaturePlot(seurat_obj_integrated,
            features = MHCmacro_genes,
            label = T,
            #split.by = "sample_id",
            #order = T,
            pt.size = 0.25,
            ncol = 6
)

VlnPlot(seurat_obj_integrated,
            features = MHCmacro_genes,
            #split.by = "sample_id",
            #order = T,
            pt.size = 0.25,
            ncol = 6
)
```


# Annotation
According to the previous cell-type identities enriched in several gene markers, we will establish the following annotation:

```{r}
# SPECIFIC
seurat_obj_integrated <- RenameIdents(
 seurat_obj_integrated,
  "0" = "Selenop+ M2-like Macrophages", # antiinflamatory
  "1" = "M1-like Macrophages", # proinflamatory
  "2" = "Slamf9+ Macrophages",
  "3" = "MHCII+Ccr2+ Macrophages", # proinflamatory
  "4" = "MHCII+Ccr2+ Macrophages", # proinflamatory
  "5" = "Monocytes",
  "6" = "Low quality cells",
  "7" = "Resident M2-like Macrophages", # antiinflamatory
  "8" = "Mature Monocytes",
  "9" = "Doublets Macro-T cells",
  "10" = "Monocyte-derived DC"
)

# Creating new annotation variable
seurat_obj_integrated$celltypes_2 <- seurat_obj_integrated@active.ident
seurat_obj_integrated$celltypes_2 <- factor(seurat_obj_integrated$celltypes_2,
                                                   levels = sort(levels(seurat_obj_integrated$celltypes_2)))

# Exploring number of celltypes
table(seurat_obj_integrated$celltypes_2)
```


## UMAP celltypes
```{r fig.width=14, fig.height=10}
# Clustering overview
p1 <- DimPlot(object = seurat_obj_integrated,
        group.by = "celltypes_2",
        label = TRUE,
        label.size = 5,
        repel = T,
        cols = color_palette
        ) +
  labs(title = "Monocyte-Macrophages")

p1
ggsave(filename = paste0(path_r_figs, "/post12_cd45pos_monomacro_clustering_annotation_umap.png"),
       plot = p1)
```

Now, we will also explore clusters across sample.
```{r fig.width=25, fig.height=5}
DimPlot(
  object = seurat_obj_integrated,
  group.by = "celltypes_2",
  split.by = "sample_id",
  label = FALSE,
  cols = color_palette
  ) + NoLegend()
```

## Cell proportions
```{r fig.width=8, fig.height=10}
gg_cell_num <- dittoBarPlot(seurat_obj_integrated, 
                            "celltypes_2",
                            group.by = "sample_id",
                            scale = "count",
                            x.reorder = c(2, 5, 3, 6, 1, 4),
                            legend.show = FALSE,
                            color.panel = color_palette)

gg_cell_prop <- dittoBarPlot(seurat_obj_integrated, 
                             "celltypes_2",
                             group.by = "sample_id",
                             scale = "percent",
                             x.reorder = c(2, 5, 3, 6, 1, 4),
                             color.panel = color_palette)

gg_cell_num / gg_cell_prop
```


# Save processed Seurat objects
```{r}
# Save lineage annotations
saveRDS(seurat_obj_integrated@meta.data[c("celltypes_2")],
        paste0(path_r_objects, "/post12_cd45pos_monomacro_annotations.rds"))

# Save processed lognorm Seurat object
saveRDS(seurat_obj_integrated, paste0(path_r_objects, "/post12_cd45pos_monomacro_clustering_annotation.rds"))

#seurat_obj_integrated <- readRDS(paste0(path_r_objects, "/post12_cd45pos_monomacro_clustering_annotation.rds"))
```


# Session Info
```{r}
sessionInfo()
```
